File: 007 有序表的合并/使用之前的顺序表实现.c
#include <stdio.h>
#include <stdlib.h>

void * merge(List *listA, List *listB) {
    int lenA = length(listA);
    int lenB = length(listB);
    
    List *listC = createList(length(listA) + length(listB));
    
    int idxA = 0;
    int idxB = 0;
    
    while (idxA < lenA && idxB < lenB) {
        int valA = getvalue(listA, idxA);
        int valB = getvalue(listB, idxB);
        if (valA <= valB) {
            append(listC, valA);
            idxA++;
        } else {
            append(listC, valB);
            idxB++;
        }
    }
    
    while (idxA < lenA) {
        append(listC, getvalue(listA, idxA));
        idxA++;
    }
    
    while (idxB < lenB) {
        append(listC, getvalue(listB, idxB));
        idxB++;
    }
    return listC;
}

int main() {
    List *listA = createList(100);
    List *listB = createList(100);
    
    int arrA[4] = {21, 32, 36, 88};
    int arrB[7] = {2, 15, 37, 99, 123, 345, 432};
    
    for (int i = 0; i < 4; i++) {
        append(listA, arrA[i]);
    }
    
    for (int i = 0; i < 7; i++) {
        append(listB, arrB[i]);
    }
    
    printf("数组A：");
    traverse(listA);
    printf("数组B：");
    traverse(listB);
    
    List *listC = merge(listA, listB);
    printf("合并后数组：");
    traverse(listC);
    
    return 0;
}

File: 007 有序表的合并/使用之前的顺序表实现.swift
import Foundation

func merge(_ listA: List<Int>, _ listB: List<Int>) -> List<Int> {
    let lenA = listA.length()
    let lenB = listB.length()
    
    var listC = List<Int>(size: UInt(lenA + lenB))
    
    var idxA: UInt = 0
    var idxB: UInt = 0
    
    while idxA < lenA && idxB < lenB {
        let valA = listA.getValue(of: idxA)!
        let valB = listB.getValue(of: idxB)!
        
        if valA < valB {
            _ = listC.append(valA)
            idxA += 1
        } else {
            _ = listC.append(valB)
            idxB += 1
        }
    }
    
    while idxA < lenA {
        _ = listC.append(listA.getValue(of: idxA)!)
        idxA += 1
    }
    
    while idxB < lenB {
        _ = listC.append(listB.getValue(of: idxB)!)
        idxB += 1
    }
    return listC
}

// 测试数据
var listA = List<Int>(size: 6)
var listB = List<Int>(size: 9)

var arrA = [Int]()
var arrB = [Int]()

for _ in 0..<6 {
    arrA.append(Int.random(in: 0..<30))
}

for _ in 0..<9 {
    arrB.append(Int.random(in: 0..<90))
}

for v in arrA.sorted() {
    _ = listA.append(v)
}

for v in arrB.sorted() {
    _ = listB.append(v)
}

print("数组A：")
listA.traverse()
print("数组B：")
listB.traverse()

print("合并后数组：")
let listC = merge(listA, listB)
listC.traverse()
